{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d1960e93-8dcf-48b6-a572-2db677cef852",
   "metadata": {},
   "source": [
    "# ENCODING "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c8e1683-d2fd-48ee-8508-38d2dd653b7c",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "84a67d1b-9eb7-4420-aff0-99b2bdb6d9ba",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#%load_ext autoreload\n",
    "#%autoreload 2\n",
    "import os, sys\n",
    "\n",
    "import sys\n",
    "sys.path.append('/home/ubuntu/')\n",
    "os.chdir('/home/ubuntu/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "d30714fe-f09c-4e4e-bc6c-edf0d40961aa",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import importlib\n",
    "import src\n",
    "importlib.reload(src)\n",
    "import src\n",
    "import src.encoding as enc\n",
    "import shlex, subprocess\n",
    "import pandas as pd\n",
    "from pybedtools import BedTool\n",
    "import numpy as np\n",
    "import seaborn as sns \n",
    "import matplotlib.pyplot as plt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "bfa40b7a-39be-41bc-876a-bb97fc539469",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "datapath= \"/mnt/data/\"\n",
    "resultspath = '/mnt/data/results/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "a0611540",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#should I add the copying of the genome file in my folder to the jupyter notebook?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31a25f49-a9fa-4b47-9537-7592a9a9dce9",
   "metadata": {
    "tags": []
   },
   "source": [
    "#creation of the file containing the length of the human chromosomes\n",
    "subprocess.run(f\"mysql --user=genome --host=genome-mysql.cse.ucsc.edu -A -e \\\n",
    "                        'select chrom, size from hg38.chromInfo' \\\n",
    "                        > {datapath}hg38.genome\", shell = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f20b1a83-ee2d-4dde-8ce7-b448a4343dc1",
   "metadata": {},
   "source": [
    "# Sloping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50b4168f-be13-475d-baef-f894ae293c19",
   "metadata": {
    "tags": []
   },
   "source": [
    "#ORIGINAL ENCODE\n",
    "#sloping the whole directory to a common  length of 400 nucleotides\n",
    "#original ENCODE\n",
    "path = f'{datapath}ENCODE/processed/'\n",
    "directory1 = os.listdir(path)\n",
    "\n",
    "\n",
    "for folder in directory1:\n",
    "    #print(folder)\n",
    "    directory2 = os.listdir(f\"{path}{folder}\")\n",
    "    \n",
    "    for file in directory2 :\n",
    "        if 'peaks.crosslink.anno.filt.bed' in file :\n",
    "            subprocess.run(f\"slopBed -i {path}/{folder}/peaks.crosslink.anno.filt.bed \\\n",
    "                                    -g {datapath}hg38.genome -b 200 \\\n",
    "                                    > {path}/{folder}/peaks.crosslink.anno.filt.slop.bed\", shell= True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "954c2179-1a23-408f-bca5-82f6f8f64eaa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a1bf3c1-2443-4bdd-8877-fd633c9e2d6a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "ca666b3a-cb6a-45ce-9426-7379351d4b32",
   "metadata": {},
   "source": [
    "# Conversion to FASTA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5eb180df-d574-45dc-91c1-3727d3526e18",
   "metadata": {
    "tags": []
   },
   "source": [
    "#ORIGINAL ENCODE\n",
    "#converting all of the eCLIP sequences to fasta \n",
    "#original ENCODE\n",
    "path = f'{datapath}ENCODE/processed/'\n",
    "directory1 = os.listdir(path)\n",
    "\n",
    "for folder in directory1:\n",
    "    directory2 = os.listdir(f\"{path}{folder}\")\n",
    "    \n",
    "    for file in directory2 :\n",
    "        \n",
    "        if 'peaks.crosslink.anno.filt.bed' in file :\n",
    "            subprocess.run(f\"bedtools getfasta -fi {datapath}GRCh38.primary_assembly.genome.fa \\\n",
    "                                    -bed {path}/{folder}/peaks.crosslink.anno.filt.slop.bed -fo \\\n",
    "                                    > {path}/{folder}/peaks.crosslink.anno.filt.slop.fa.out\", shell= True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "940b8d46-94b6-44ab-8203-41d82e744c90",
   "metadata": {
    "tags": []
   },
   "source": [
    "#NEW NEGATIVES\n",
    "# converting negatives into fasta\n",
    "#New negative dataset\n",
    "\n",
    "path = f'{datapath}negatives/'\n",
    "directory1 = os.listdir(path)\n",
    "\n",
    "for folder in directory1:\n",
    "    directory2 = os.listdir(f\"{path}{folder}\")\n",
    "    \n",
    "    for file in directory2 :\n",
    "        subprocess.run(f\"bedtools getfasta -fi {datapath}GRCh38.primary_assembly.genome.fa \\\n",
    "                                        -bed {path}{folder}/{file} -fo \\\n",
    "                                        > {path}{folder}/{file}.fa\", shell= True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ecb526e-0df9-4fb7-858f-ab47289cc250",
   "metadata": {},
   "source": [
    "# Removal of files (if necessary) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "739f1fc5-ecee-44ba-bfa8-b6f6055527cf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#removal of previous files \n",
    "#modified dataset\n",
    "\n",
    "path = f'{datapath}processed/ENCODE-Sofia/'\n",
    "directory1 = os.listdir(path)\n",
    "\n",
    "\n",
    "for folder in directory1:\n",
    "    #if folder == 'RBM15_HepG2':\n",
    "    current_folder = folder\n",
    "    directory2 = os.listdir(f\"{path}{folder}\")\n",
    "\n",
    "    for folder2 in directory2 :\n",
    "        directory3 = os.listdir(f\"{path}{folder}/{folder2}\")\n",
    "\n",
    "        for file in directory3 :\n",
    "\n",
    "            if 'out.df' in file :\n",
    "                subprocess.run(f\"rm {path}{folder}/{folder2}/{file}\", shell= True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db7a10a8-54e0-4c8c-a1ac-7d08a7f6a0bb",
   "metadata": {
    "tags": []
   },
   "source": [
    "# preparing folders for plots \n",
    "\n",
    "path = f'{datapath}processed/ENCODE-Sofia/'\n",
    "directory1 = os.listdir(path)\n",
    "\n",
    "df_methylation = pd.read_csv(f'{resultspath}methylation_rate.csv')\n",
    "df_high = df_methylation[df_methylation.methylation == 'high']\n",
    "list_high_meth = list(df_high.RBP)\n",
    "\n",
    "# tic()\n",
    "\n",
    "for folder in directory1:\n",
    "    if folder not in list_high_meth:\n",
    "        current_folder = folder\n",
    "        directory2 = os.listdir(f\"{path}{folder}\")\n",
    "        subprocess.run(f\"rmdir results/plots/summary_metrics/{folder}\", shell= True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76294684-0b7e-41fa-956a-3169da95be47",
   "metadata": {
    "tags": []
   },
   "source": [
    "#removal of previous files \n",
    "#modified dataset\n",
    "\n",
    "path = f'{datapath}negatives/'\n",
    "directory1 = os.listdir(path)\n",
    "\n",
    "\n",
    "for folder in directory1:\n",
    "    current_folder = folder\n",
    "    directory3 = os.listdir(f\"{path}{folder}\")\n",
    "\n",
    "    for file in directory3 :\n",
    "\n",
    "        if '.miclip' in file :\n",
    "            subprocess.run(f\"rm {path}{folder}/{file}\", shell= True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "052e5a1a-8d46-4b42-981d-82d81ae5598f",
   "metadata": {},
   "source": [
    "# m6A sites filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a5b220f-7c63-4145-8fe3-b2605f45e2cc",
   "metadata": {
    "tags": []
   },
   "source": [
    "#NEW NEGATIVES\n",
    "#m6A sites filtering for new set of negatives \n",
    "\n",
    "\n",
    "path = f'{datapath}negatives/'\n",
    "directory1 = os.listdir(path)\n",
    "\n",
    "for folder in directory1: \n",
    "    current_folder = folder\n",
    "    directory2 = os.listdir(path + '/' + folder)\n",
    "\n",
    "    for file in directory2 :\n",
    "\n",
    "        if '.bed.fa' not in file :\n",
    "            subprocess.run(f\"bedtools intersect -wb \\\n",
    "                                    -a {datapath}miCLIP.filt.out.bed \\\n",
    "                                    -b {path}{folder}/{file} \\\n",
    "                                     > {path}{folder}/{file}.miclip.filt.bed.out\", shell= True)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45d1b262-e4a8-49f9-bf90-93fe6a874faf",
   "metadata": {
    "tags": []
   },
   "source": [
    "#CV DATASET\n",
    "#m6A sites filtering for pos, neg-1, neg-2 and folds --> for one hot encoding --> isn't the fasta missing? no, it's included in the cv dataset already\n",
    "\n",
    "\n",
    "path = f'{datapath}processed/ENCODE-Sofia/'\n",
    "directory1 = os.listdir(path)\n",
    "\n",
    "for folder in directory1:\n",
    "    current_folder = folder\n",
    "    directory2 = os.listdir(f\"{path}{folder}\")\n",
    "\n",
    "    for folder2 in directory2 :\n",
    "        directory3 = os.listdir(f\"{path}{folder}/{folder2}\")\n",
    "\n",
    "        for file in directory3 :\n",
    "\n",
    "            if '.bed' in file :\n",
    "                subprocess.run(f\"bedtools intersect -wb \\\n",
    "                                        -a {datapath}miCLIP.filt.out.bed \\\n",
    "                                        -b {path}/{folder}/{folder2}/{file} \\\n",
    "                                        > {path}/{folder}/{folder2}/{file}.miclip.filt.bed.out\", shell= True)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "670115e7-c66a-457e-943c-6ce040f7d023",
   "metadata": {},
   "source": [
    "# DATA AUGMENTATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b79e174-1ca2-4302-866d-d67ecf626c81",
   "metadata": {},
   "source": [
    "# just CV dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5a4fd89-efb8-427d-9114-d34ea81f80b4",
   "metadata": {
    "tags": []
   },
   "source": [
    "#first retrieving the RPB binding site from .bed\n",
    "path = f'{datapath}processed/ENCODE-Sofia/'\n",
    "directory1 = os.listdir(path)\n",
    "\n",
    "\n",
    "for folder in directory1:  \n",
    "    current_folder = folder\n",
    "    directory2 = os.listdir(f\"{path}{folder}\")\n",
    "\n",
    "    for folder2 in directory2 :\n",
    "        directory3 = os.listdir(f\"{path}{folder}/{folder2}\")\n",
    "\n",
    "        for file in directory3 :\n",
    "            #this is just for positives, since negative 1 don't have to contain any RBP site  \n",
    "            if ('.bed' in file ) and  ('out' not in file) :\n",
    "                subprocess.run(f\"slopBed -i {path}{folder}/{folder2}/{file} \\\n",
    "                                    -g {datapath}hg38.genome -l -200 -r -200 \\\n",
    "                                    > {path}{folder}/{folder2}/{file}.out\", shell= True)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ae6bb3c-44ed-44d4-8ee8-9e91b21252de",
   "metadata": {
    "tags": []
   },
   "source": [
    "#now I'm going to slop the coordinates randomly --> bed.out.slop.out\n",
    "\n",
    "path = f'{datapath}processed/ENCODE-Sofia/'\n",
    "directory1 = os.listdir(path)\n",
    "\n",
    "\n",
    "for folder in directory1:   \n",
    "    current_folder = folder\n",
    "    directory2 = os.listdir(f\"{path}{folder}\")\n",
    "\n",
    "    for folder2 in directory2 :\n",
    "        directory3 = os.listdir(f\"{path}{folder}/{folder2}\")\n",
    "\n",
    "        for file in directory3 :\n",
    "        #considering the fact that negative 1 do not have to contain any RBP binding site\n",
    "            \n",
    "            if ('.bed.out' in file ) and  ('slop' not in file) and ('miclip' not in file):\n",
    "                l,r = enc.random_l_r(399,2) #-l and -r should be randomized and their sum should be 400. \n",
    "                subprocess.run(f\"slopBed -i {path}{folder}/{folder2}/{file} \\\n",
    "                                        -g {datapath}hg38.genome -l {l} -r {r} \\\n",
    "                                        > {path}{folder}/{folder2}/{file}.slop.out\", shell= True)\n",
    "\n",
    "            \n",
    "\n",
    "            if ('.bed.out' in file ) and ( 'negative' in file ) and  ('slop' not in file) and ('miclip' not in file):\n",
    "                for i in [0, 1, 2]:\n",
    "                    l,r = enc.random_l_r(399,2)            \n",
    "                    subprocess.run(f\"slopBed -i {path}{folder}/{folder2}/{file} \\\n",
    "                                            -g {datapath}hg38.genome -l {l} -r {r} \\\n",
    "                                            > {path}{folder}/{folder2}/{file}.slop.out{i}\", shell= True)\n",
    "                    subprocess.run(f'cat {path}{folder}/{folder2}/{file}.slop.out{i} \\\n",
    "                                        >> {path}{folder}/{folder2}/{file}.slop.out', shell = True)  #here merge the files into {path}/{folder}/{folder2}/{file}.slop.out\n",
    "                    subprocess.run(f'rm {path}{folder}/{folder2}/{file}.slop.out{i}', shell = True)\n",
    "                    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03223dc1-fcbd-486d-a8b8-ea4371833293",
   "metadata": {
    "tags": []
   },
   "source": [
    "#now I have the slopped sequences, I'll check if the m6A site is included -- the file is very similar to miclip.filt.bed.out , the code is same\n",
    "#later to filter out out.slop.out.fasta file --> out.slop.out.miclip.filt.bed.out contains \n",
    "\n",
    "path = f'{datapath}processed/ENCODE-Sofia/'\n",
    "directory1 = os.listdir(path)\n",
    "\n",
    "for folder in directory1:\n",
    "    current_folder = folder\n",
    "    directory2 = os.listdir(f\"{path}{folder}\")\n",
    "\n",
    "    for folder2 in directory2 :\n",
    "        directory3 = os.listdir(f\"{path}{folder}/{folder2}\")\n",
    "\n",
    "        for file in directory3 :\n",
    "\n",
    "            if ('.out.slop.out' in file) and ('fasta' not in file) and ('miclip' not in file) :\n",
    "                subprocess.run(f\"bedtools intersect -wb \\\n",
    "                                        -a {datapath}miCLIP.filt.out.bed \\\n",
    "                                        -b {path}{folder}/{folder2}/{file} \\\n",
    "                                       > {path}{folder}/{folder2}/{file}.miclip.filt.bed.out\", shell= True)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61d74007-f62b-42e2-8e1b-ff187feacb26",
   "metadata": {
    "tags": []
   },
   "source": [
    "#now I have to convert the new sequences (out.slop.out)  into fasta\n",
    "\n",
    "path = f'{datapath}processed/ENCODE-Sofia/'\n",
    "directory1 = os.listdir(path)\n",
    "\n",
    "\n",
    "for folder in directory1:\n",
    "    current_folder = folder\n",
    "    directory2 = os.listdir(f\"{path}{folder}\")\n",
    "\n",
    "    for folder2 in directory2 :\n",
    "        directory3 = os.listdir(f\"{path}{folder}/{folder2}\")\n",
    "\n",
    "        for file in directory3 :\n",
    "\n",
    "            if ('out.slop.out' in file) and ('.miclip.filt.bed' not in file) and ('fasta' not in file) :\n",
    "                subprocess.run(f\"bedtools getfasta -fi {datapath}GRCh38.primary_assembly.genome.fa \\\n",
    "                                            -bed {path}{folder}/{folder2}/{file} -fo {path}{folder}/{folder2}/{file}.fasta\", shell= True)\n",
    "\n",
    "                   \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "189bb486-26e0-4b4a-b9fc-4ea88dddc27f",
   "metadata": {},
   "source": [
    "# New Augmentation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "39c30c5e-552f-4d88-9cc0-0ba5cceea37b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "# define function to be used as lambda in the apply() later on\n",
    "def get_center_pos_between_rbp_and_m6a(x0, x1):\n",
    "    m6a_pos = x0 # this should be single-nucleotide so doesn't matter if you take start or end\n",
    "    rbp_pos = x1\n",
    "    dist = rbp_pos - m6a_pos # this is a positive quantity if m6a_pos left, negative otherwise\n",
    "    return int( rbp_pos + dist / 2)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "1d105415-3189-4a7d-b60d-d9a9576027d4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "positive.fold-4.bed.miclip.filt.bed.out\n",
      "     chrom  pos_between_rbp_and_m6a  end_pos_between_rbp_and_m6a chrom_m6a   \n",
      "0    chr10                110590432                    110590432     chr10  \\\n",
      "1    chr10                110590430                    110590430     chr10   \n",
      "2    chr10                110601766                    110601766     chr10   \n",
      "3    chr10                110601732                    110601732     chr10   \n",
      "4    chr10                 17229514                     17229514     chr10   \n",
      "..     ...                      ...                          ...       ...   \n",
      "732   chrX                135033052                    135033052      chrX   \n",
      "733   chrX                135033043                    135033043      chrX   \n",
      "734   chrX                135033064                    135033064      chrX   \n",
      "735   chrX                 44524182                     44524182      chrX   \n",
      "736   chrX                 72272593                     72272593      chrX   \n",
      "\n",
      "     start_m6a    end_m6a                                 details  score   \n",
      "0    110590430  110590431   B045_1-4_chr10_f_c846[k=4][m=2]_AGACA      1  \\\n",
      "1    110590430  110590431   B045_1-4_chr10_f_c846[k=4][m=2]_AGACA      1   \n",
      "2    110601754  110601755   B045_1-4_chr10_f_c848[k=4][m=2]_GAACT      1   \n",
      "3    110601754  110601755   B045_1-4_chr10_f_c848[k=4][m=2]_GAACT      1   \n",
      "4     17229515   17229516  B045_1-4_chr10_f_c124[k=21][m=2]_TGACT      1   \n",
      "..         ...        ...                                     ...    ...   \n",
      "732  135032989  135032990   B045_1-4_chrX_f_c679[k=13][m=2]_GTACA      1   \n",
      "733  135032997  135032998   B045_1-4_chrX_f_c680[k=15][m=2]_GGACG      1   \n",
      "734  135032997  135032998   B045_1-4_chrX_f_c680[k=15][m=2]_GGACG      1   \n",
      "735   44524209   44524210    B045_1-4_chrX_r_c179[k=6][m=2]_GGACT      1   \n",
      "736   72272653   72272654   B045_1-4_chrX_r_c334[k=81][m=2]_TGACA      1   \n",
      "\n",
      "    strand chrom_gene  ...  thickStart  thickEnd  a chrom_rbp  start_rbp   \n",
      "0        +      chr10  ...           .         +  1     chr10  110590225  \\\n",
      "1        +      chr10  ...           .         +  1     chr10  110590230   \n",
      "2        +      chr10  ...           .         +  1     chr10  110601529   \n",
      "3        +      chr10  ...           .         +  1     chr10  110601598   \n",
      "4        +      chr10  ...           .         +  1     chr10   17229317   \n",
      "..     ...        ...  ...         ...       ... ..       ...        ...   \n",
      "732      +       chrX  ...           .         +  1      chrX  135032662   \n",
      "733      +       chrX  ...           .         +  1      chrX  135032705   \n",
      "734      +       chrX  ...           .         +  1      chrX  135032662   \n",
      "735      -       chrX  ...           .         -  1      chrX   44524062   \n",
      "736      -       chrX  ...           .         -  1      chrX   72272572   \n",
      "\n",
      "       end_rbp                b         c  d     bs_pos  \n",
      "0    110590625  PUM1_K562_rep01  3.712086  +  110590425  \n",
      "1    110590630  PUM1_K562_rep02  2.861600  +  110590430  \n",
      "2    110601929  PUM1_K562_rep01  2.823979  +  110601729  \n",
      "3    110601998  PUM1_K562_rep01  2.590779  +  110601798  \n",
      "4     17229717  PUM1_K562_rep02  3.584759  +   17229517  \n",
      "..         ...              ...       ... ..        ...  \n",
      "732  135033062  PUM1_K562_rep01  2.565244  +  135032862  \n",
      "733  135033105  PUM1_K562_rep02  3.106712  +  135032905  \n",
      "734  135033062  PUM1_K562_rep01  2.565244  +  135032862  \n",
      "735   44524462  PUM1_K562_rep01  2.632359  -   44524262  \n",
      "736   72272972  PUM1_K562_rep01  2.546385  -   72272772  \n",
      "\n",
      "[737 rows x 23 columns]\n",
      "negative-2.fold-4.bed.miclip.filt.bed.out\n",
      "     chrom  pos_between_rbp_and_m6a  end_pos_between_rbp_and_m6a chrom_m6a   \n",
      "0    chr10                100275330                    100275330     chr10  \\\n",
      "1    chr10                119577068                    119577068     chr10   \n",
      "2    chr10                119898650                    119898650     chr10   \n",
      "3    chr10                 96521083                     96521083     chr10   \n",
      "4    chr10                 97333542                     97333542     chr10   \n",
      "..     ...                      ...                          ...       ...   \n",
      "284   chr9                136831344                    136831344      chr9   \n",
      "285   chr9                136831310                    136831310      chr9   \n",
      "286   chr9                 19376295                     19376295      chr9   \n",
      "287   chr9                 19376623                     19376623      chr9   \n",
      "288   chrX                 48528486                     48528486      chrX   \n",
      "\n",
      "     start_m6a    end_m6a                                 details  score   \n",
      "0    100275257  100275258  B045_1-4_chr10_r_c657[k=10][m=2]_TCACA      1  \\\n",
      "1    119577096  119577097   B045_1-4_chr10_r_c861[k=7][m=2]_TGACT      1   \n",
      "2    119898589  119898590   B045_1-4_chr10_f_c925[k=6][m=2]_GGACA      1   \n",
      "3     96521100   96521101   B045_1-4_chr10_r_c616[k=9][m=2]_CTACA      1   \n",
      "4     97333578   97333579  B045_1-4_chr10_r_c620[k=21][m=2]_CTACT      1   \n",
      "..         ...        ...                                     ...    ...   \n",
      "284  136831371  136831372    B045_1-4_chr9_f_c858[k=4][m=2]_GGACA      1   \n",
      "285  136831371  136831372    B045_1-4_chr9_f_c858[k=4][m=2]_GGACA      1   \n",
      "286   19376364   19376365    B045_1-4_chr9_r_c89[k=31][m=2]_GAACA      1   \n",
      "287   19376583   19376584    B045_1-4_chr9_r_c90[k=43][m=2]_AAACG      1   \n",
      "288   48528478   48528479    B045_1-4_chrX_f_c237[k=4][m=2]_GAACA      1   \n",
      "\n",
      "    strand chrom_gene  ...  thickStart  thickEnd  a chrom_rbp  start_rbp   \n",
      "0        -      chr10  ...           .         -  1     chr10  100274911  \\\n",
      "1        -      chr10  ...           .         -  1     chr10  119576952   \n",
      "2        +      chr10  ...           .         +  1     chr10  119898267   \n",
      "3        -      chr10  ...           .         -  1     chr10   96520934   \n",
      "4        -      chr10  ...           .         -  1     chr10   97333450   \n",
      "..     ...        ...  ...         ...       ... ..       ...        ...   \n",
      "284      +       chr9  ...           .         +  1      chr9  136831224   \n",
      "285      +       chr9  ...           .         +  1      chr9  136831293   \n",
      "286      -       chr9  ...           .         -  1      chr9   19376302   \n",
      "287      -       chr9  ...           .         -  1      chr9   19376302   \n",
      "288      +       chrX  ...           .         +  1      chrX   48528262   \n",
      "\n",
      "       end_rbp                    b         c  d     bs_pos  \n",
      "0    100275311     TIA1_HepG2_rep02  2.724557  -  100275111  \n",
      "1    119577352   BCLAF1_HepG2_rep01  4.435767  -  119577152  \n",
      "2    119898667   IGF2BP1_K562_rep01  5.668302  +  119898467  \n",
      "3     96521334  FAM120A_HepG2_rep01  5.792964  -   96521134  \n",
      "4     97333850    RBM15_HepG2_rep02  2.591565  -   97333650  \n",
      "..         ...                  ...       ... ..        ...  \n",
      "284  136831624     RBM15_K562_rep01  4.057813  +  136831424  \n",
      "285  136831693     AGGF1_K562_rep01  3.484820  +  136831493  \n",
      "286   19376702    LARP4_HepG2_rep01  3.325905  -   19376502  \n",
      "287   19376702    LARP4_HepG2_rep01  3.325905  -   19376502  \n",
      "288   48528662     RPS3_HepG2_rep01  3.928770  +   48528462  \n",
      "\n",
      "[289 rows x 23 columns]\n",
      "negative-1.fold-4.bed.miclip.filt.bed.out\n",
      "    chrom  pos_between_rbp_and_m6a  end_pos_between_rbp_and_m6a chrom_m6a   \n",
      "0   chr10                 17234776                     17234776     chr10  \\\n",
      "1   chr11                 34100324                     34100324     chr11   \n",
      "2   chr11                 47779081                     47779081     chr11   \n",
      "3   chr11                 47779142                     47779142     chr11   \n",
      "4   chr11                 47779153                     47779153     chr11   \n",
      "5    chr1                150578894                    150578894      chr1   \n",
      "6   chr11                 65079661                     65079661     chr11   \n",
      "7   chr11                 65079678                     65079678     chr11   \n",
      "8   chr11                  9139654                      9139654     chr11   \n",
      "9    chr1                212619024                    212619024      chr1   \n",
      "10   chr1                225402580                    225402580      chr1   \n",
      "11   chr1                244862456                    244862456      chr1   \n",
      "12   chr1                244862678                    244862678      chr1   \n",
      "13   chr1                 32685908                     32685908      chr1   \n",
      "14  chr13                 48261180                     48261180     chr13   \n",
      "15  chr13                 48261248                     48261248     chr13   \n",
      "16  chr15                 41766578                     41766578     chr15   \n",
      "17  chr15                 58897195                     58897195     chr15   \n",
      "18  chr15                 93027450                     93027450     chr15   \n",
      "19  chr15                 93027860                     93027860     chr15   \n",
      "20  chr16                 19704707                     19704707     chr16   \n",
      "21  chr16                 30084710                     30084710     chr16   \n",
      "22  chr16                 30084719                     30084719     chr16   \n",
      "23  chr16                 30084854                     30084854     chr16   \n",
      "24  chr16                 30085033                     30085033     chr16   \n",
      "25  chr16                 30085108                     30085108     chr16   \n",
      "26  chr17                 10692769                     10692769     chr17   \n",
      "27  chr17                 40161976                     40161976     chr17   \n",
      "28  chr17                 48596105                     48596105     chr17   \n",
      "29   chr1                  8870408                      8870408      chr1   \n",
      "30  chr21                 43761652                     43761652     chr21   \n",
      "31   chr2                 47801141                     47801141      chr2   \n",
      "32   chr2                 47801208                     47801208      chr2   \n",
      "33   chr3                139358716                    139358716      chr3   \n",
      "34   chr3                 42638811                     42638811      chr3   \n",
      "35   chr3                 50288050                     50288050      chr3   \n",
      "36   chr3                 50288246                     50288246      chr3   \n",
      "37   chr6                159753601                    159753601      chr6   \n",
      "38   chr6                 30722644                     30722644      chr6   \n",
      "39   chr7                 45108958                     45108958      chr7   \n",
      "40   chrX                 81114125                     81114125      chrX   \n",
      "\n",
      "    start_m6a    end_m6a                                 details  score   \n",
      "0    17234723   17234724  B045_1-4_chr10_f_c130[k=14][m=4]_GAACA      1  \\\n",
      "1    34100295   34100296   B045_1-4_chr11_f_c316[k=6][m=2]_TAACA      1   \n",
      "2    47779010   47779011  B045_1-4_chr11_r_c408[k=12][m=2]_TAACT      1   \n",
      "3    47779051   47779052  B045_1-4_chr11_r_c409[k=25][m=4]_AGACT      1   \n",
      "4    47779058   47779059   B045_1-4_chr11_r_c410[k=8][m=2]_GGACT      1   \n",
      "5   150578946  150578947   B045_1-4_chr1_r_c1518[k=6][m=2]_GGACA      1   \n",
      "6    65079717   65079718  B045_1-4_chr11_r_c616[k=13][m=2]_AGACA      1   \n",
      "7    65079728   65079729   B045_1-4_chr11_r_c617[k=7][m=2]_GGACC      1   \n",
      "8     9139644    9139645   B045_1-4_chr11_r_c161[k=4][m=2]_GGACT      1   \n",
      "9   212619001  212619002   B045_1-4_chr1_f_c1953[k=4][m=2]_TGACT      1   \n",
      "10  225402609  225402610  B045_1-4_chr1_r_c2244[k=12][m=2]_TAACT      1   \n",
      "11  244862494  244862495   B045_1-4_chr1_r_c2536[k=7][m=2]_GAACA      1   \n",
      "12  244862642  244862643  B045_1-4_chr1_r_c2537[k=25][m=2]_GTACA      1   \n",
      "13   32685880   32685881    B045_1-4_chr1_f_c575[k=4][m=2]_TAACT      1   \n",
      "14   48261224   48261225  B045_1-4_chr13_f_c182[k=14][m=4]_GAACA      1   \n",
      "15   48261269   48261270   B045_1-4_chr13_f_c183[k=8][m=2]_TAACC      1   \n",
      "16   41766551   41766552   B045_1-4_chr15_f_c133[k=4][m=2]_TGACT      1   \n",
      "17   58897171   58897172   B045_1-4_chr15_r_c232[k=4][m=2]_GGACT      1   \n",
      "18   93027543   93027544   B045_1-4_chr15_f_c605[k=6][m=2]_AGACA      1   \n",
      "19   93027816   93027817   B045_1-4_chr15_f_c607[k=4][m=2]_GGACT      1   \n",
      "20   19704796   19704797  B045_1-4_chr16_r_c310[k=16][m=6]_GAACT      1   \n",
      "21   30084749   30084750  B045_1-4_chr16_f_c545[k=13][m=2]_TGACA      1   \n",
      "22   30084755   30084756  B045_1-4_chr16_f_c546[k=11][m=2]_TGACA      1   \n",
      "23   30084845   30084846   B045_1-4_chr16_f_c547[k=4][m=2]_CTACT      1   \n",
      "24   30084964   30084965  B045_1-4_chr16_f_c548[k=15][m=3]_GGACG      1   \n",
      "25   30085014   30085015  B045_1-4_chr16_f_c549[k=16][m=2]_AGACA      1   \n",
      "26   10692864   10692865   B045_1-4_chr17_r_c290[k=6][m=2]_GGACT      1   \n",
      "27   40162023   40162024   B045_1-4_chr17_f_c518[k=4][m=2]_GAACA      1   \n",
      "28   48596140   48596141   B045_1-4_chr17_r_c783[k=4][m=2]_GAACT      1   \n",
      "29    8870477    8870478   B045_1-4_chr1_r_c230[k=33][m=5]_AAACT      1   \n",
      "30   43761654   43761655   B045_1-4_chr21_f_c305[k=6][m=2]_TGACT      1   \n",
      "31   47801074   47801075   B045_1-4_chr2_f_c372[k=28][m=4]_GGACT      1   \n",
      "32   47801119   47801120    B045_1-4_chr2_f_c373[k=5][m=2]_GGACT      1   \n",
      "33  139358774  139358775   B045_1-4_chr3_r_c1076[k=4][m=2]_AGACT      1   \n",
      "34   42638897   42638898    B045_1-4_chr3_f_c332[k=4][m=2]_GGACC      1   \n",
      "35   50288036   50288037   B045_1-4_chr3_r_c458[k=15][m=2]_GGACA      1   \n",
      "36   50288167   50288168   B045_1-4_chr3_r_c459[k=12][m=2]_AGACT      1   \n",
      "37  159753584  159753585   B045_1-4_chr6_f_c1235[k=4][m=2]_ATACA      1   \n",
      "38   30722612   30722613    B045_1-4_chr6_f_c340[k=6][m=2]_GGACC      1   \n",
      "39   45108871   45108872    B045_1-4_chr7_r_c428[k=8][m=2]_ATACA      1   \n",
      "40   81114079   81114080    B045_1-4_chrX_r_c551[k=8][m=2]_AGACT      1   \n",
      "\n",
      "   strand chrom_gene  ...  thickStart  thickEnd  a chrom_rbp  start_rbp   \n",
      "0       +      chr10  ...           .         +  1     chr10   17234417  \\\n",
      "1       +      chr11  ...           .         +  1     chr11   34100037   \n",
      "2       -      chr11  ...           .         -  1     chr11   47778668   \n",
      "3       -      chr11  ...           .         -  1     chr11   47778668   \n",
      "4       -      chr11  ...           .         -  1     chr11   47778668   \n",
      "5       -       chr1  ...           .         -  1      chr1  150578849   \n",
      "6       -      chr11  ...           .         -  1     chr11   65079628   \n",
      "7       -      chr11  ...           .         -  1     chr11   65079628   \n",
      "8       -      chr11  ...           .         -  1     chr11    9139423   \n",
      "9       +       chr1  ...           .         +  1      chr1  212618754   \n",
      "10      -       chr1  ...           .         -  1      chr1  225402467   \n",
      "11      -       chr1  ...           .         -  1      chr1  244862370   \n",
      "12      -       chr1  ...           .         -  1      chr1  244862370   \n",
      "13      +       chr1  ...           .         +  1      chr1   32685623   \n",
      "14      +      chr13  ...           .         +  1     chr13   48261111   \n",
      "15      +      chr13  ...           .         +  1     chr13   48261111   \n",
      "16      +      chr15  ...           .         +  1     chr15   41766297   \n",
      "17      -      chr15  ...           .         -  1     chr15   58896923   \n",
      "18      +      chr15  ...           .         +  1     chr15   93027528   \n",
      "19      +      chr15  ...           .         +  1     chr15   93027528   \n",
      "20      -      chr16  ...           .         -  1     chr16   19704773   \n",
      "21      +      chr16  ...           .         +  1     chr16   30084626   \n",
      "22      +      chr16  ...           .         +  1     chr16   30084626   \n",
      "23      +      chr16  ...           .         +  1     chr16   30084626   \n",
      "24      +      chr16  ...           .         +  1     chr16   30084626   \n",
      "25      +      chr16  ...           .         +  1     chr16   30084626   \n",
      "26      -      chr17  ...           .         -  1     chr17   10692854   \n",
      "27      +      chr17  ...           .         +  1     chr17   40161917   \n",
      "28      -      chr17  ...           .         -  1     chr17   48596010   \n",
      "29      -       chr1  ...           .         -  1      chr1    8870415   \n",
      "30      +      chr21  ...           .         +  1     chr21   43761457   \n",
      "31      +       chr2  ...           .         +  1      chr2   47800740   \n",
      "32      +       chr2  ...           .         +  1      chr2   47800740   \n",
      "33      -       chr3  ...           .         -  1      chr3  139358690   \n",
      "34      +       chr3  ...           .         +  1      chr3   42638868   \n",
      "35      -       chr3  ...           .         -  1      chr3   50287808   \n",
      "36      -       chr3  ...           .         -  1      chr3   50287808   \n",
      "37      +       chr6  ...           .         +  1      chr6  159753350   \n",
      "38      +       chr6  ...           .         +  1      chr6   30722347   \n",
      "39      -       chr7  ...           .         -  1      chr7   45108497   \n",
      "40      -       chrX  ...           .         -  1      chrX   81113786   \n",
      "\n",
      "      end_rbp                                                  b    c    d   \n",
      "0    17234817  ENSG00000026025.16;ENSG00000026025.16;ENSG0000...    .    +  \\\n",
      "1    34100437  ENSG00000135387.21;ENSG00000135387.21;ENSG0000...    .    +   \n",
      "2    47779068  ENSG00000030066.14;ENSG00000030066.14;ENSG0000...    .    -   \n",
      "3    47779068  ENSG00000030066.14;ENSG00000030066.14;ENSG0000...    .    -   \n",
      "4    47779068  ENSG00000030066.14;ENSG00000030066.14;ENSG0000...    .    -   \n",
      "5   150579249  ENSG00000203804.5;ENSG00000143384.14;ENSG00000...    .    -   \n",
      "6    65080028  ENSG00000146670.10;ENSG00000146670.10;ENSG0000...    .    -   \n",
      "7    65080028  ENSG00000146670.10;ENSG00000146670.10;ENSG0000...    .    -   \n",
      "8     9139823  ENSG00000184014.9;ENSG00000184014.9;ENSG000001...    .    -   \n",
      "9   212619154  ENSG00000162772.17;ENSG00000162772.17;ENSG0000...    .    +   \n",
      "10  225402867  ENSG00000143815.15;ENSG00000143815.15;ENSG0000...    .    -   \n",
      "11  244862770  ENSG00000153187.20;ENSG00000153187.20;ENSG0000...    .    -   \n",
      "12  244862770  ENSG00000153187.20;ENSG00000153187.20;ENSG0000...    .    -   \n",
      "13   32686023  ENSG00000162521.19;ENSG00000162521.19;ENSG0000...    .    +   \n",
      "14   48261511  ENSG00000136156.15;ENSG00000136156.15;ENSG0000...    .    +   \n",
      "15   48261511  ENSG00000136156.15;ENSG00000136156.15;ENSG0000...    .    +   \n",
      "16   41766697  ENSG00000174197.18;ENSG00000174197.18;ENSG0000...    .    +   \n",
      "17   58897323  ENSG00000137776.17;ENSG00000137776.17;ENSG0000...  NaN  NaN   \n",
      "18   93027928  ENSG00000272888.8;ENSG00000272888.8;ENSG000002...  NaN  NaN   \n",
      "19   93027928  ENSG00000272888.8;ENSG00000272888.8;ENSG000002...  NaN  NaN   \n",
      "20   19705173  ENSG00000103550.14;ENSG00000103550.14;ENSG0000...    .    -   \n",
      "21   30085026  ENSG00000149923.14;ENSG00000149923.14;ENSG0000...    .    +   \n",
      "22   30085026  ENSG00000149923.14;ENSG00000149923.14;ENSG0000...    .    +   \n",
      "23   30085026  ENSG00000149923.14;ENSG00000149923.14;ENSG0000...    .    +   \n",
      "24   30085026  ENSG00000149923.14;ENSG00000149923.14;ENSG0000...    .    +   \n",
      "25   30085026  ENSG00000149923.14;ENSG00000149923.14;ENSG0000...    .    +   \n",
      "26   10693254  ENSG00000133028.12;ENSG00000133028.12;ENSG0000...    .    -   \n",
      "27   40162317  ENSG00000108349.17;ENSG00000108349.17;ENSG0000...    .    +   \n",
      "28   48596410  ENSG00000120093.12;ENSG00000120093.12;ENSG0000...    .    -   \n",
      "29    8870815  ENSG00000074800.16;ENSG00000074800.16;ENSG0000...    .    -   \n",
      "30   43761857  ENSG00000160209.19;ENSG00000160209.19;ENSG0000...    .    +   \n",
      "31   47801140  ENSG00000116062.18;ENSG00000116062.18;ENSG0000...    .    +   \n",
      "32   47801140  ENSG00000116062.18;ENSG00000116062.18;ENSG0000...    .    +   \n",
      "33  139359090  ENSG00000184432.11;ENSG00000184432.11;ENSG0000...    .    -   \n",
      "34   42639268  ENSG00000114857.19;ENSG00000114857.19;ENSG0000...    .    +   \n",
      "35   50288208  ENSG00000214706.12;ENSG00000214706.12;ENSG0000...    .    -   \n",
      "36   50288208  ENSG00000214706.12;ENSG00000214706.12;ENSG0000...    .    -   \n",
      "37  159753750  ENSG00000146457.16;ENSG00000146457.16;ENSG0000...    .    +   \n",
      "38   30722747  ENSG00000196230.14;ENSG00000196230.14;ENSG0000...    .    +   \n",
      "39   45108897  ENSG00000136270.14;ENSG00000136270.14;ENSG0000...    .    -   \n",
      "40   81114186  ENSG00000198157.11;ENSG00000198157.11;ENSG0000...    .    -   \n",
      "\n",
      "       bs_pos  \n",
      "0    17234617  \n",
      "1    34100237  \n",
      "2    47778868  \n",
      "3    47778868  \n",
      "4    47778868  \n",
      "5   150579049  \n",
      "6    65079828  \n",
      "7    65079828  \n",
      "8     9139623  \n",
      "9   212618954  \n",
      "10  225402667  \n",
      "11  244862570  \n",
      "12  244862570  \n",
      "13   32685823  \n",
      "14   48261311  \n",
      "15   48261311  \n",
      "16   41766497  \n",
      "17   58897123  \n",
      "18   93027728  \n",
      "19   93027728  \n",
      "20   19704973  \n",
      "21   30084826  \n",
      "22   30084826  \n",
      "23   30084826  \n",
      "24   30084826  \n",
      "25   30084826  \n",
      "26   10693054  \n",
      "27   40162117  \n",
      "28   48596210  \n",
      "29    8870615  \n",
      "30   43761657  \n",
      "31   47800940  \n",
      "32   47800940  \n",
      "33  139358890  \n",
      "34   42639068  \n",
      "35   50288008  \n",
      "36   50288008  \n",
      "37  159753550  \n",
      "38   30722547  \n",
      "39   45108697  \n",
      "40   81113986  \n",
      "\n",
      "[41 rows x 23 columns]\n"
     ]
    }
   ],
   "source": [
    "path = f'{datapath}processed/ENCODE-Sofia/'\n",
    "directory1 = os.listdir(path)\n",
    "%env CUDA_DEVICE_ORDER=PCI_BUS_ID\n",
    "%env CUDA_VISIBLE_DEVICES=0\n",
    "df_methylation = pd.read_csv(f'{resultspath}methylation_rate.csv')\n",
    "df_high = df_methylation[df_methylation.methylation == 'high']\n",
    "list_high_meth = list(df_high.RBP)\n",
    "\n",
    "\n",
    "for folder in directory1:\n",
    "    if folder in list_high_meth:\n",
    "    #if folder == 'RBM15_HepG2':\n",
    "        current_folder = folder\n",
    "        directory2 = os.listdir(f\"{path}{folder}\")\n",
    "\n",
    "        for folder2 in directory2 :\n",
    "            directory3 = os.listdir(f\"{path}{folder}/{folder2}\")\n",
    "\n",
    "            for file in directory3 :\n",
    "                if ('miclip.filt.bed.out' in file) and ('df' not in file): \n",
    "                    try: \n",
    "                        bed = BedTool(f'{path}{folder}/{folder2}/{file}')\n",
    "                        df = bed.to_dataframe(disable_auto_names=True, names = [\"chrom_m6a\", \"start_m6a\", \"end_m6a\", \"details\",'score', 'strand',\"chrom_gene\",\"start_gene\",\"end_gene\",\"ids\", 'thickStart', 'thickEnd' ,'a', \"chrom_rbp\", \"start_rbp\", \"end_rbp\", 'b', 'c','d'])\n",
    "                        #df = pd.DataFrame(f'{path}{folder}/{folder2}/{file}', sep=\"\\t\", names=[\"chrom_m6a\", \"start_m6a\", \"end_m6a\", \"details\",\"chrom_gene\",\"start_gene\",\"end_gene\", \"ids\" ,\"chrom_rbp\", \"start_rbp\", \"end_rbp\"])\n",
    "                        # get original bindingsite position\n",
    "                        df['bs_pos'] = df['start_rbp'].apply(lambda x: x + 200)\n",
    "                        df['pos_between_rbp_and_m6a'] = df[['bs_pos', 'start_m6a']].apply(lambda x: enc.get_center_pos_between_rbp_and_m6a(x[0], x[1]), axis=1)\n",
    "                        df['chrom'] = df['chrom_m6a']\n",
    "                        df['end_pos_between_rbp_and_m6a']= df['pos_between_rbp_and_m6a']\n",
    "                        cols = df.columns.tolist()\n",
    "                        cols = ['chrom','pos_between_rbp_and_m6a','end_pos_between_rbp_and_m6a',\"chrom_m6a\", \"start_m6a\", \"end_m6a\", \"details\",'score', 'strand',\"chrom_gene\",\"start_gene\",\"end_gene\",\"ids\", 'thickStart', 'thickEnd' ,'a', \"chrom_rbp\", \"start_rbp\", \"end_rbp\", 'b', 'c','d','bs_pos']\n",
    "                        df = df[cols]\n",
    "                        #print(df)\n",
    "                        BedTool.from_dataframe(df,f'{path}{folder}/{folder2}/{file}.df' )\n",
    "                    except: \n",
    "                        print(f'error occurred in {path}{folder}/{folder2}/{file}')\n",
    "                        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "71450b8b-a1bd-4f2e-b83c-1c7830ae8db2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#now I'm going to slop the coordinates randomly --> bed.out.slop.out\n",
    "%env CUDA_DEVICE_ORDER=PCI_BUS_ID\n",
    "%env CUDA_VISIBLE_DEVICES=0\n",
    "path = f'{datapath}processed/ENCODE-Sofia/'\n",
    "directory1 = os.listdir(path)\n",
    "\n",
    "df_methylation = pd.read_csv(f'{resultspath}methylation_rate.csv')\n",
    "df_high = df_methylation[df_methylation.methylation == 'high']\n",
    "list_high_meth = list(df_high.RBP)\n",
    "\n",
    "\n",
    "for folder in directory1:\n",
    "    if folder in list_high_meth: \n",
    "    #if folder == 'RBM15_HepG2':\n",
    "        current_folder = folder\n",
    "        directory2 = os.listdir(f\"{path}{folder}\")\n",
    "\n",
    "        for folder2 in directory2 :\n",
    "            directory3 = os.listdir(f\"{path}{folder}/{folder2}\")\n",
    "\n",
    "            for file in directory3 :\n",
    "                #considering the fact that negative 1 do not have to contain any RBP binding site\n",
    "\n",
    "                if ('miclip.filt.bed.out.df' in file ) and ('positive' in file) and ('slop' not in file):\n",
    "                    l,r = enc.random_l_r(400,2) #-l and -r should be randomized and their sum should be 400. \n",
    "                    subprocess.run(f\"slopBed -i {path}{folder}/{folder2}/{file} \\\n",
    "                                            -g {datapath}hg38.genome -l {l} -r {r} \\\n",
    "                                            > {path}{folder}/{folder2}/{file}.slop\", shell= True)\n",
    "\n",
    "\n",
    "\n",
    "                if ('miclip.filt.bed.out.df' in file ) and ('negative' in file) and ('slop' not in file):\n",
    "                    for i in range(30):\n",
    "                        l,r = enc.random_l_r(400,2)            \n",
    "                        subprocess.run(f\"slopBed -i {path}{folder}/{folder2}/{file} \\\n",
    "                                                -g {datapath}hg38.genome -l {l} -r {r} \\\n",
    "                                                > {path}{folder}/{folder2}/{file}.slop{i}\", shell= True)\n",
    "                        subprocess.run(f'cat {path}{folder}/{folder2}/{file}.slop{i} \\\n",
    "                                            >> {path}{folder}/{folder2}/{file}.slop', shell = True)  #here merge the files into {path}/{folder}/{folder2}/{file}.slop.out\n",
    "                        subprocess.run(f'rm {path}{folder}/{folder2}/{file}.slop{i}', shell = True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "2bdc7700-5e05-4281-aad1-af4341d3548d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#now I have to convert the new sequences (out.slop.out)  into fasta\n",
    "%env CUDA_DEVICE_ORDER=PCI_BUS_ID\n",
    "%env CUDA_VISIBLE_DEVICES=0\n",
    "path = f'{datapath}processed/ENCODE-Sofia/'\n",
    "directory1 = os.listdir(path)\n",
    "\n",
    "df_methylation = pd.read_csv(f'{resultspath}methylation_rate.csv')\n",
    "df_high = df_methylation[df_methylation.methylation == 'high']\n",
    "list_high_meth = list(df_high.RBP)\n",
    "\n",
    "\n",
    "for folder in directory1:\n",
    "    if folder in list_high_meth: \n",
    "    #if folder == 'RBM15_HepG2':\n",
    "        current_folder = folder\n",
    "        directory2 = os.listdir(f\"{path}{folder}\")\n",
    "\n",
    "        for folder2 in directory2 :\n",
    "            directory3 = os.listdir(f\"{path}{folder}/{folder2}\")\n",
    "\n",
    "            for file in directory3 :\n",
    "\n",
    "                if ('miclip.filt.bed.out.df.slop' in file) and ('fasta' not in file) :\n",
    "                    subprocess.run(f\"bedtools getfasta -fi {datapath}GRCh38.primary_assembly.genome.fa \\\n",
    "                                                -bed {path}{folder}/{folder2}/{file} -fo {path}{folder}/{folder2}/{file}.fasta\", shell= True)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "5f9ba6da-726e-4818-b317-763ff1064783",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#now I have the slopped sequences, I'll check if the m6A site is included -- the file is very similar to miclip.filt.bed.out , the code is same\n",
    "#later to filter out out.slop.out.fasta file --> out.slop.out.miclip.filt.bed.out contains \n",
    "%env CUDA_DEVICE_ORDER=PCI_BUS_ID\n",
    "%env CUDA_VISIBLE_DEVICES=0\n",
    "path = f'{datapath}processed/ENCODE-Sofia/'\n",
    "directory1 = os.listdir(path)\n",
    "\n",
    "df_methylation = pd.read_csv(f'{resultspath}methylation_rate.csv')\n",
    "df_high = df_methylation[df_methylation.methylation == 'high']\n",
    "list_high_meth = list(df_high.RBP)\n",
    "\n",
    "\n",
    "for folder in directory1:\n",
    "    if folder in list_high_meth: \n",
    "    #if folder == 'RBM15_HepG2': \n",
    "        current_folder = folder\n",
    "        directory2 = os.listdir(f\"{path}{folder}\")\n",
    "\n",
    "        for folder2 in directory2 :\n",
    "            directory3 = os.listdir(f\"{path}{folder}/{folder2}\")\n",
    "\n",
    "            for file in directory3 :\n",
    "\n",
    "                if ('miclip.filt.bed.out.df.slop' in file) and ('fasta' not in file) :\n",
    "                    subprocess.run(f\"bedtools intersect -wb \\\n",
    "                                            -a {datapath}miCLIP.filt.out.bed \\\n",
    "                                            -b {path}{folder}/{folder2}/{file} \\\n",
    "                                           > {path}{folder}/{folder2}/{file}.miclip.filt.bed.out\", shell= True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efea11c4-1201-4f4c-b9b8-56a1e57a7f46",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e037b0fc-01a2-4391-aaa4-32abc21c0656",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51366e23-6f96-4fa1-9f7e-c142df20405f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
